### PR TITLE
[SYCL][ESIMD][EMU] single_task support, v.2

### PR DESCRIPTION
**Pull request by lsatanov**
- ESIMD EMU: single_task() support.
- ESIMD plugin: misc clarifying refactoring.

### CODE REVIEW

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp [hunk 1]
```diff
@@ -16,6 +16,7 @@
 
 #include <CL/sycl/backend_types.hpp>
 #include <CL/sycl/detail/accessor_impl.hpp>
+#include <CL/sycl/detail/cg_types.hpp> // for HostKernelBase
```
Comment (lsatanov) [inline sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:19]: Remnant from variant 1. Will eliminate after testing completes

**Review by dongkyunahn-intel — DISMISSED** [review]

File: sycl/include/CL/sycl/handler.hpp [hunk 2]
```diff
@@ -660,16 +660,29 @@ class __SYCL_EXPORT handler {
...
+    return ResetHostKernelHelper<KernelType, struct NormalizedKernelType, Dims>(
+        KernelFunc);
+  }
+
+  /* 'wrapper'-based approach using 'NormalizedKernelType' struct is not used
+   * for 'void(sycl::group<Dims>)' since 'void(sycl::group<Dims>)' is not
+   * supported in ESIMD.
    */
-  // For 'void' and 'sycl::group<Dims>' kernel argument
+  // For 'sycl::group<Dims>' kernel argument
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/handler.hpp:683]: please use single comment style - '//'

File: sycl/include/CL/sycl/detail/cg_types.hpp [hunk 3]
```diff
@@ -246,6 +246,7 @@ template <class KernelType, class KernelArgType, int Dims>
 class HostKernel : public HostKernelBase {
   using IDBuilder = sycl::detail::Builder;
   KernelType MKernel;
+
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/detail/cg_types.hpp:249]: here and in other places - please avoid unrelated changes like adding empty lines

File: sycl/include/CL/sycl/handler.hpp [hunk 4]
```diff
@@ -660,16 +660,29 @@ class __SYCL_EXPORT handler {
...
+        detail::runKernelWithoutArg(MKernelFunc);
+      }
+    };
+    return ResetHostKernelHelper<KernelType, struct NormalizedKernelType, Dims>(
+        KernelFunc);
+  }
+
+  /* 'wrapper'-based approach using 'NormalizedKernelType' struct is not used
+   * for 'void(sycl::group<Dims>)' since 'void(sycl::group<Dims>)' is not
+   * supported in ESIMD.
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/handler.hpp:681]: Is `sycl::group<Dims>` kernel supported at all (non-ESIMD case)? If not, then RT error would be better.

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp [hunk 5]
```diff
@@ -147,33 +147,15 @@ using KernelFunc = std::function<void(const sycl::nd_item<NDims> &)>;
 
 // Struct to wrap dimension info and lambda function to be invoked by
 // CM Kernel launcher that only accepts raw function pointer for
-// kernel execution. Function instances of 'InvokeLambda' un-wrap this
-// struct instance and invoke lambda function ('Func')
-template <int NDims> struct LambdaWrapper {
+// kernel execution. Function instances of 'KernelWrapper' un-wrap
+// this struct instance and invoke lambda function ('Func')
+template <int NDims> struct KernelWrapperData {
```
Comment (kbobrovs) [inline sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:152]: I don't see how KernelWrapper is better than InvokeLambda. If Lambda does not cover all kernel cases then InvokeKernel or KernelInvoker?
Then KernelWrapperData -> KernelInvocationContext.

If this all is only supposed to be used on a host device (and/or esimd emulator), then I'd add 'Host' prefix as well.

**Review by kbobrovs — COMMENTED** [review]
@romanovvlad, @smaslov-intel - please take a look as well.

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp (hunk 5 above)
Comment (lsatanov) [reply sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:152]: Warning: naming discussion.
Ok, the point was to make more sense (and less code). Yes, kernel is not always lambda.

There are many ways to do it. I consider KernelWrapper and KernelWrapperData clear enough and find adding "Host" prefix for things used inside esimd_emulator only an overkill and not bringing more sense.

If inevitable, the naming can be changed **to the required liking**.
e.g.,

InvokeKernel
KernelInvokationContext

would be totally clear.

File: sycl/include/CL/sycl/handler.hpp (hunk 4 above)
Comment (lsatanov) [reply sycl/include/CL/sycl/handler.hpp:681]: My current understanding is the following: it's a "won't happen" case for ESIMD, hence no error.
Shall review this.

**Review by romanovvlad — DISMISSED** [review]
LGTM

**Review by dongkyunahn-intel — APPROVED** [review]

**Comment by lsatanov** [comment]
@romanovvlad, @kbobrovs hi. Merging?

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp (hunk 5 above)
Comment (kbobrovs) [reply sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp]: Classes should not be named with verbs, but adhjectives. KernelCaller, KernelInvoker

> Warning: naming discussion.
Not sure I understand the warning. You started changing names, so I commented.

File: sycl/include/CL/sycl/handler.hpp (hunk 4 above)
Comment (kbobrovs) [reply sycl/include/CL/sycl/handler.hpp]: Assuming something won't happen and just ignoring the possibility is not safe. Please add error reporting.

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp (hunk 5 above)
Comment (lsatanov) [reply sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp]: Konstantin, InvokeKernel is a function (template), not a class.

P.S. regarding classes, how they should be named (verbs/adjectives/nouns) is a debatable taking into account that they can be callable in C++ (as we all know). In other languages, e.g. JavaScript, everything is an object. Should we name everything in JavaScript not using verbs? I doubt that.

**Review by romanovvlad — APPROVED** [review]

File: sycl/include/CL/sycl/handler.hpp (hunk 4 above)
Comment (lsatanov) [reply sycl/include/CL/sycl/handler.hpp]: That's true. This code was introduced previously and results in runtime failure in case of using esimd emu backend.
This is to be addressed in the following changes.

**Comment by bader** [comment]
@lsatanov, this patch break the build due to 
```
/home/runner/work/llvm/llvm/src/sycl/include/CL/sycl/handler.hpp:671:44: error: unused parameter 'Arg' [-Werror,-Wunused-parameter]
      void operator()(const nd_item<Dims> &Arg) {
                                           ^
1 error generated.
```
Could you fix this issue, please?
See full log here: https://github.com/intel/llvm/runs/5493988722?check_suite_focus=true

**Comment by lsatanov** [comment]
> @lsatanov, this patch break the build due to
> 
> ```
> /home/runner/work/llvm/llvm/src/sycl/include/CL/sycl/handler.hpp:671:44: error: unused parameter 'Arg' [-Werror,-Wunused-parameter]
>       void operator()(const nd_item<Dims> &Arg) {
>                                            ^
> 1 error generated.
> ```
> 
> Could you fix this issue, please? See full log here: https://github.com/intel/llvm/runs/5493988722?check_suite_focus=true

Sure! #5779

Interesting, how the pre-commit checks have passed (?)

**Comment by bader** [comment]
Post-commit has wider coverage. Thanks for fixing it!
